okhttp/src/commonJvmAndroid/kotlin/okhttp3/internal/http/CallServerInterceptor.kt

```diff
@@ -125,14 +125,27 @@ class CallServerInterceptor(
       exchange.responseHeadersEnd(response)
 
       response =
-        if (forWebSocket && code == 101) {
+        if (forWebSocket && code == HTTP_SWITCHING_PROTOCOLS) {
           // Connection is upgrading, but we need to ensure interceptors see a non-null response body.
           response.stripBody()
         } else {
-          response
-            .newBuilder()
-            .body(exchange.openResponseBody(response))
-            .build()
+          if (code == HTTP_SWITCHING_PROTOCOLS &&
+            "upgrade".equals(response.request.header("Connection"), ignoreCase = true) &&
+            "upgrade".equals(response.header("Connection"), ignoreCase = true) &&
+            "tcp".equals(response.request.header("Upgrade"), ignoreCase = true) &&
```

Comment:
The PR title feels like it's about all possible protocols, but in practice seems tied to solving TCP specifically for docker?

Is there a spec for this TCP upgrade?

Reply:
You're right, I wasn't aware of the more general possibilities for a connection upgrade, and the case for Websockets hasn't changed with my pull request.

My current aim is in fact to add support for TCP upgrades, because my use case is driven by the Docker engine api.
There's no spec for the TCP upgrade per se, at least not known to me.

Given the issue at https://github.com/square/okhttp/issues/5874 I would suggest to start with support for `tcp` and `websocket`. Other protocols might be added in the future.

Reply:
I actually think we might make things easy for ourselves by omitting the `tcp` check here. We can handle all upgrades this way, and let the caller decide what they want to do with their socket based on the contents of the `Upgrade` header.

Reply:
I removed the `tcp` check as you suggested. I also changed the connection upgrade check to make the similarities between Websocket and generic upgrades more visible. This might help with the next steps/discussion.